drivers/hwinfo/hwinfo_sam.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2019 Aurelien Jarno
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <hwinfo.h>
+#include <init.h>
+#include <soc.h>
+#include <string.h>
+
+static u8_t sam_uid[16];
+
+ssize_t z_impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
+{
+	if (length > sizeof(sam_uid)) {
+		length = sizeof(sam_uid);
+	}
+
+	memcpy(buffer, sam_uid, length);
+
+	return length;
+}
+
+/* On the Atmel SAM SoC series, the device id is located in the flash
+ * controller. The controller can either present the flash area containing
+ * the code, the unique identifier or the user signature area at the flash
+ * location. Therefore the function reading the device id must be executed
+ * from RAM with the interrupts disabled. To avoid executing this complex
+ * code each time the device id is requested, we do this at boot time at save
+ * the 128-bit value into RAM.
+ */
+__ramfunc static void hwinfo_sam_read_device_id(void)
+{
+	Efc *efc = (Efc *)DT_FLASH_DEV_BASE_ADDRESS;
+	u8_t *flash = (u8_t *)CONFIG_FLASH_BASE_ADDRESS;
```

Comment:
One more question here, @aurel32 : *flash points to the beginning of physical flash, not the image start. Is this intentional?

Reply:
>The Unique Identifier is located in the first 128 bits of the Flash memory mapping. So, at the address 0x80000-0x80003.
According to
http://asf.atmel.com/docs/latest/sam.services.flash.read.unique.id_example.sam3x_ek/html/index.html

Reply:
ok